marimba/core/schemas/ifdo.py

```diff
@@ -0,0 +1,341 @@
+"""
+Marimba iFDO Metadata Implementation.
+
+This module provides functionality for handling iFDO metadata, including creation, processing, and embedding of
+metadata in image files. It implements the BaseMetadata interface for iFDO-specific metadata and offers methods
+for creating dataset metadata and processing image files with EXIF data.
+
+Imports:
+    io: Provides tools for working with I/O operations
+    json: Handles JSON data encoding and decoding
+    datetime: Supplies classes for working with dates and times
+    Fraction: Represents rational numbers with numerator and denominator
+    Path: Offers object-oriented filesystem paths
+    typing: Provides support for type hints
+    uuid: Generates universally unique identifiers
+    piexif: Handles reading and writing of EXIF data in images
+    PIL: Python Imaging Library for opening, manipulating, and saving image files
+    rich: Offers rich text and beautiful formatting in the terminal
+
+Classes:
+    iFDOMetadata: Implements the BaseMetadata interface for iFDO-specific metadata
+"""
+
+import io
+import json
+from datetime import datetime, timezone
+from fractions import Fraction
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, cast
+from uuid import uuid4
+
+import piexif
+from PIL import Image
+from rich.progress import Progress, SpinnerColumn, TaskID
+
+from marimba.core.schemas.base import BaseMetadata
+from marimba.core.utils.log import get_logger
+from marimba.core.utils.rich import get_default_columns
+from marimba.lib import image
+from marimba.lib.decorators import multithreaded
+from marimba.lib.gps import convert_degrees_to_gps_coordinate
+
+if TYPE_CHECKING:
+    from ifdo.models import ImageData, ImageSetHeader, iFDO
+else:
+    from ifdo.models import ImageData, ImageSetHeader, iFDO
+
+
+logger = get_logger(__name__)
+
+
+class iFDOMetadata(BaseMetadata):  # noqa: N801
```

Comment:
As I noted on email, this class might be better suited to handle a list of `ImageData` as opposed to a single item. Most of this code that adapts `ImageData` to the requirements of `ImageData` would likely still work by using the first element of the list. See https://marine-imaging.com/fair/ifdos/iFDO-overview/#ifdo-parts

That way, pipeline authors would include a single `iFDOMetadata` instance in the list of metadata instances for a given file. The `iFDOMetadata` would encapsulate the list of `ImageData` items (one for an image, potentially multiple for a video) as well as the logic to support the hooks that `BaseMetadata` requires.